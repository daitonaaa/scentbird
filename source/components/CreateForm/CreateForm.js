import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import './CreateForm.scss';

import {
  createItem,
  createChild
} from 'actions/items';


export class CreateForm extends Component {

  static propTypes = {
    parentId: PropTypes.number, // только если childPosition = true
    childPosition: PropTypes.bool,

    list: PropTypes.array.isRequired,

    createItem: PropTypes.func.isRequired,
    createChild: PropTypes.func.isRequired,
  };

  state = {
    error: '',
    titleValue: '',
  }

  handleChangeTitle = event => {
    const state = {
      error: '',
      titleValue: event.target.value,
    };

    this.setState({ ...state });
  }

  handleCreateItem = () => {
    const { titleValue } = this.state;
    const {
      list,
      parentId,
      createItem,
      createChild,
      childPosition,
    } = this.props;

    if (titleValue && list instanceof Array) {

      if (childPosition) {
        const parentIndex = list.findIndex(
          item => item.id === parentId
        );

        if (!(parentIndex + 1)) {
          this.setState({ error: 'Родитель не найден' });

          return;
        }

        const childs = list[parentIndex].childs;
        const newId = (
          childs.length
            ? childs[childs.length - 1].id + 1
            : 1
        );

        createChild(parentId, titleValue, newId);
        this.setState({
          titleValue: '',
          error: '',
        });
      }

      else {
        const newId = (
          list.length
            ? list[list.length - 1].id + 1
            : 1
        );

        createItem(titleValue, newId);
        this.setState({
          titleValue: '',
          error: '',
        });
      }
    }

    else this.setState({
      error: 'Вы должны ввести название элемента',
    });
  }

  renderError() {
    const { error } = this.state;

    if (error) return (
      <div className="create-item__form-error">
        {error}
      </div>
    );
  }

  render() {
    const { titleValue } = this.state;

    return (
      <div className="create-item__form">
        <div className="create-item__form-box">
          <input
            value={titleValue}
            onChange={this.handleChangeTitle}
            className="create-item__form-input"
            placeholder="Название нового элемента"
          />
          <div
            onClick={this.handleCreateItem}
            className="create-item__form-add"
          >
            Добавить новый элемент в список
          </div>
        </div>
        {this.renderError()}
      </div>
    );
  }
}


const mapStateToProps = state => ({
  list: state.items.list,
});

const mapDispatchToProps = {
  createItem,
  createChild,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateForm);
