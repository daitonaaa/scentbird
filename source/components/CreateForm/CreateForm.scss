.create-item__form {
  position: relative;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 15px;

  &-input {
    padding: 5px;
    width: 250px;
    background: #e9f2ff;
    color: #4c5163;
    text-align: center;
    border: solid 1px #c8cfe8;
    border-radius: 5px;
    transition: all ease .2s;

    &:focus {
      border: solid 1px #7c9fb5;
    }
  }

  &-add {
    display: inline-block;
    margin-left: 15px;
    padding: 5px 15px;
    background: #7c9fb5;
    color: #fff;
    border-radius: 5px;
    cursor: pointer;
    transition: all ease .2s;

    &:hover {
      background: #70a7ca;
    }
  }

  &-error {
    margin-top: 10px;
    width: 100%;
    font-size: 12px;
    color: #c53b3b;
    text-align: center;
  }
}
