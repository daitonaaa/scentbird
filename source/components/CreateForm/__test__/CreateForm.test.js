import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { CreateForm } from 'components/CreateForm/CreateForm';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });


const defaultProps = {
  list: [],
  createItem: ()=>{},
  createChild: ()=>{},
};


describe('> > > COMPONENT - CreateForm', () => {
  it('#create snapshot', () => {
    const output = shallow(
      <CreateForm {...defaultProps} />
    );

    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('#test method: handleCreateItem => no title error', () => {
    const output = shallow(
      <CreateForm {...defaultProps} />
    );

    output.find('.create-item__form-add').simulate('click');
    expect(output.state('error')).toBe('Вы должны ввести название элемента');
    expect(output.instance().renderError()).toBeDefined();
  });

  it('#test method: handleCreateItem => no parent error', () => {
    const output = shallow(
      <CreateForm childPosition {...defaultProps} />
    );

    output.find('.create-item__form-input').simulate(
      'change', { target: { value: 'title' } }
    );
    output.find('.create-item__form-add').simulate('click');
    expect(output.state('error')).toBe('Родитель не найден');
  });

  it('#test method: handleCreateItem => success create Item', () => {
    defaultProps.list = [
      { id: 1 }, { id: 2 },
    ];
    const output = shallow(
      <CreateForm {...defaultProps} />
    );

    output.find('.create-item__form-input').simulate(
      'change', { target: { value: 'title' } }
    );
    expect(output.find('.create-item__form-input').prop('value')).toBe('title');

    output.find('.create-item__form-add').simulate('click');
    expect(output.state('titleValue')).toBe('');
  });

  it('#test method: handleCreateItem => success create Child', () => {
    defaultProps.list = [
      { id: 1 },
      { id: 2, childs: [
        { id: 1 }
      ] },
    ];
    const output = shallow(
      <CreateForm parentId={2} childPosition {...defaultProps} />
    );

    output.find('.create-item__form-input').simulate(
      'change', { target: { value: 'title' } }
    );
    output.find('.create-item__form-add').simulate('click');
    expect(output.state('titleValue')).toBe('');
  });
});
