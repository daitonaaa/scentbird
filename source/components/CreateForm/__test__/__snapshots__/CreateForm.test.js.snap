// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`> > > COMPONENT - CreateForm #create snapshot 1`] = `
<div
  className="create-item__form"
>
  <div
    className="create-item__form-box"
  >
    <input
      className="create-item__form-input"
      onChange={[Function]}
      placeholder="Название нового элемента"
      value=""
    />
    <div
      className="create-item__form-add"
      onClick={[Function]}
    >
      Добавить новый элемент в список
    </div>
  </div>
</div>
`;
